### ISSUE TITLE
[OC 0.6.6, HP EliteBook 840 G4)]PS2 Trackpoint and SMBus Touchpad - how to get them both to function properly

### ISSUE THREAD

**Issue by XXCoreRangerX** [leo-labs/VoodooSMBus#66]
I have made a [custom EFI](https://github.com/XXCoreRangerX/Hackintosh-HP-EliteBook-840-G4) for the HP EliteBook 840 G4. I managed to get the touchpad to work using `VoodooRMI`, but am struggling to fix the trackpoint. It does work if I use `VoodooPS2Mouse`, but it breaks the SMBus touchpad and in this repo's installation guide it's recommended to be disabled. `VoodooRMI` alone doesn't fix the trackpoint, as it's PS2 based and won't work with SMBus. How can I make the trackpoint work?

**Comment by averycblack** [comment]
Would you mind joining the Gitter? Also in the future, please post VoodooRMI issues in the VoodooRMI repo.

Atleast with Lenovo devices, normally the trackpoint operates under PS2 passthrough under the function F03, not sure why that would not be the case here with your HP. Can you please provide a log using [`sudo dmesg`](https://github.com/VoodooSMBus/VoodooRMI#troubleshooting) as listed here, using the latest version of VoodooRMI DEBUG?

**Comment by XXCoreRangerX** [comment]
[Here's output from dmesg.](https://github.com/VoodooSMBus/VoodooSMBus/files/6143366/vrmi-dmesg.log)

It looks like it doesn't use F03 at all. Do I need to provide more logs?

**Comment by averycblack** [comment]
Make sure that you have `msgbuf=...` in your boot-args as listed in the link, and don't use the trackpad until you get the logs. I need the init data, which will get overwritten if you use the trackpad.

**Comment by XXCoreRangerX** [comment]
Alright, these are the actual [logs](https://github.com/VoodooSMBus/VoodooSMBus/files/6143489/dmesg-output.log). It still looks like F03 is not being used.

**Comment by averycblack** [comment]
Hrmm, yeah you're right. What kext do you use if using the trackpad under PS2? Does the trackpoint work there?
I'd be interested to see what linux or windows would show if at all possible.

**Comment by XXCoreRangerX** [comment]
Yeah, trackpoint works using VoodooPS2Mouse from Acidanthera's VoodooPS2Controller. Trackpad has very bad support using that kext, and works like a regular mouse. I couldn't use VoodooPS2Trackpad and VoodooPS2Mouse or VoodooPS2Mouse and VoodooRMI at the same time without breaking the trackpoint.

On windows the trackpoint is named Synaptics PS2 Device (or something like that).

This is what IOReg shows about the trackpoint and VoodooRMI (without VoodooPS2Mouse enabled):
![screenshot-1](https://user-images.githubusercontent.com/61242573/111200377-f7f2d100-85c1-11eb-84db-a4203329a0e0.png)
![screenshot-2](https://user-images.githubusercontent.com/61242573/111200358-ef01ff80-85c1-11eb-8562-d2f531b79347.png)

PS2M is mentioned in the DSDT but there's not much info about it.
```asl
        Device (PS2M)
        {
            Name (_CID, Package (0x03)  // _CID: Compatible ID
            {
                EisaId ("SYN0100"), 
                EisaId ("SYN0002"), 
                EisaId ("PNP0F13") /* PS/2 Mouse */
            })
            Name (_CRS, ResourceTemplate ()  // _CRS: Current Resource Settings
            {
                IRQ (Edge, ActiveHigh, Exclusive, )
                    {12}
            })
            Name (_PRS, ResourceTemplate ()  // _PRS: Possible Resource Settings
            {
                StartDependentFn (0x00, 0x00)
                {
                    IRQNoFlags ()
                        {12}
                }
                EndDependentFn ()
            })
        }
```

**Comment by XXCoreRangerX** [comment]
This is what IOReg shows when I enable VoodooPS2Mouse:
![screenshot-3](https://user-images.githubusercontent.com/61242573/111202238-ed393b80-85c3-11eb-95a1-1329fb718fad.png)
![screenshot-4](https://user-images.githubusercontent.com/61242573/111202923-b9aae100-85c4-11eb-811d-99bef21b2e57.png)

By enabling that kext I can use the trackpoint and even the top touchpad buttons (these are dedicated for the trackpoint). However, VoodooRMI fails to load:
```
[    8.920569]: VRMI - Error: Error: Failed to read SMBus version. Code: 0xFFFFFFFA
```
But when I do `kextutil showloaded` - I can see that both VoodooRMI and RMISMBus are "loaded".

**P.S.** VoodooPS2 has a cool double tap to drag with touchpad implementation. Would be cool if that would be added to VoodooRMI as well, because it's really useful. VoodooPS2 also lights up the keyboard when the touchpad or trackpoint is used.

**Comment by XXCoreRangerX** [comment]
What can I do to have both the trackpoint and trackpad working?

**Comment by averycblack** [comment]
Hrmm - is it reasonable for you to boot linux? Even if it's just a live USB?
I'm not really sure how your trackpoint is supposed to work...

If you can, please run `cat /proc/bus/input/devices` and send the output here? Can redirect it to a file with ` > file path` and send the file.

**Comment by XXCoreRangerX** [comment]
Sure, give me a few minutes.

**Comment by XXCoreRangerX** [comment]
Here's the [output](https://github.com/VoodooSMBus/VoodooSMBus/files/6144070/devices.log). All input devices including the trackpoint worked on Linux.

**Comment by XXCoreRangerX** [comment]
Pretty sure the trackpoint is the device named `PS/2 Generic Mouse` in the log.

**Comment by averycblack** [comment]
Hrmm - I'm going to go out on a limb and say this would probably require some extensive work on the PS2 side to make this work. It appears to be on the Mux port (Not Keyboard or Aux port), with no passthrough available. You're right that it shows up as a generic PS2 mouse. This unfortunately is not related to VoodooRMI really, it's just a side product of not using the trackpad over PS2 anymore.

Just to double check, can you please run `sudo dmesg | grep serio` and `sudo dmesg | grep input` in Linux?

**Comment by XXCoreRangerX** [comment]
[Output](https://github.com/VoodooSMBus/VoodooSMBus/files/6144160/dmesg-input-serio.log)

Is there any way to make it compatible with VoodooRMI tho?

**Comment by averycblack** [comment]
I can't do anything on my side with VoodooRMI - it is possible to make the trackpoint work while using VoodooRMI, but it requires work on the PS2 side like I said. Some way to differentiate and only config the trackpoint is needed. If you mess with the trackpad at all over PS2 (without resetting it), it makes it impossible to initialize over SMBus. So like I said, requires work over PS2.

Looking at you're log, I'm guessing that some sort of mux is in place, or multiple AUX ports are being used. I'm guessing that `KBD port` is Serio0, `AUX0` is Serio1 (which contains the trackpad), and then `AUX1-3` is Serio2-4. All the aux ports use the same irq and are all represented by the PS2M ACPI object that you showed above. If you can figure out how to address just AUX1, then you should be able to get it work over PS2

https://github.com/torvalds/linux/blob/master/drivers/input/serio/i8042.c#L1352-L1356
https://github.com/torvalds/linux/blob/master/drivers/input/serio/i8042.c#L1455-L1467

To be clear, this isn't any sort of pass through occuring through the trackpad - these are two distinct devices on different PS2 ports. This is why you don't have F03 (PS2 passthrough)

Edit: Your best bet is to likely talk to Acidanthera as they maintain VoodooPS2 currently, or roll your own PS2 driver.
Edit 2: More sources about how interrupts work with multiple ports it seems: https://github.com/torvalds/linux/blob/master/drivers/input/serio/i8042.c#L518

**Comment by XXCoreRangerX** [comment]
Alright, thanks for your time. Appreciate your help. I'll keep working on this.

**Comment by averycblack** [comment]
https://github.com/torvalds/linux/blob/master/drivers/input/serio/i8042.c#L542-L579
So, in the interrupt, [VoodooPS2](https://github.com/acidanthera/VoodooPS2/blob/master/VoodooPS2Controller/VoodooPS2Controller.cpp#L174-L180) does get this data - so it looks to be that it shouldn't be *that* difficult to add mux support. I don't think VoodooPS2Trackpad would need much, if any adaptation, and same with Mouse/Trackpad. It's all changes within VoodooPS2Controller.

**Comment by zhen-zen** [comment]
Please report available RMI functions shown in debug log and try https://github.com/VoodooSMBus/VoodooRMI/pull/113.

**Comment by XXCoreRangerX** [comment]
Hello! These are RMI functions that appear in dmesg:
```
[    7.685066]: VRMI - Debug: rmi_probe_interrupts: Counting IRQs
[    7.710505]: VRMI - Debug: IRQ Count: 6
[    7.716336]: VRMI - Debug: rmi_init_functions: Creating functions
[    7.724519]: VRMI - Info:  Initializing F34.
[    7.730134]: VRMI - Debug: Function F34 - IRQs: 1 CMD Base: 0 CTRL Base: 61 DATA Base: 0 QRY Base: 112 VER: 0
[    7.737319]: VRMI - Info:  F34 not implemented
[    7.744969]: VRMI - Info:  Initializing F01.
[    7.750581]: VRMI - Debug: Function F1 - IRQs: 1 CMD Base: 110 CTRL Base: 65 DATA Base: 19 QRY Base: 121 VER: 0
[    7.773500]: VRMI - Info:  Found RMI4 device, manufacturer: Synaptics, product: TM3139-001, fw id: 2000742
[    7.788506]: VRMI - Info:  Initializing F11.
[    7.794224]: VRMI - Debug: Function F11 - IRQs: 2 CMD Base: 111 CTRL Base: 71 DATA Base: 21 QRY Base: 148 VER: 1
[    7.828317]: VRMI - Info:  Initializing F54.
[    7.834079]: VRMI - Debug: Function F54 - IRQs: 1 CMD Base: 353 CTRL Base: 270 DATA Base: 256 QRY Base: 354 VER: 0
[    7.841684]: VRMI - Info:  F54 not implemented
[    7.849524]: VRMI - Info:  Initializing F30.
[    7.855453]: VRMI - Debug: Function F30 - IRQs: 1 CMD Base: 256 CTRL Base: 341 DATA Base: 269 QRY Base: 368 VER: 0
[    7.867294]: VRMI - Debug: F30: Found Button 1
[    7.873356]: VRMI - Debug: F30: Found Button 2
```

The only functions that are implemented are F01, F11 and F30
![image](https://user-images.githubusercontent.com/61242573/118403676-cfde2780-b66f-11eb-9d09-41e71afa9d07.png)

I built the kext in Xcode (from your pull request branch) and placed in EFI/OC/Kexts. The trackpoint and upper buttons are still not working.

**Comment by zhen-zen** [comment]
Could be F54 but the function seems to be an image device... I'm not sure how to verify it in Linux.
~https://android.googlesource.com/kernel/tegra/+/android-7.1.1_r0.79/drivers/input/touchscreen/rmi4/rmi_f54.c~
https://elixir.bootlin.com/linux/latest/source/drivers/input/rmi4/rmi_f54.c

**Comment by averycblack** [comment]
> 
> 
> Could be F54 but the function seems to be an image device... I'm not sure how to verify it in Linux.
> ~https://android.googlesource.com/kernel/tegra/+/android-7.1.1_r0.79/drivers/input/touchscreen/rmi4/rmi_f54.c~
> https://elixir.bootlin.com/linux/latest/source/drivers/input/rmi4/rmi_f54.c

I doubt it's F54 - my trackpad has that function as well in addition to F03. I'm not seeing any trackpoint related code as well

**Comment by XXCoreRangerX** [comment]
On Linux, there's an `lsirq` command. It displays kernel interrupts information.

**TL;DR:** RMI functions active on Linux: `FN01`, `FN11`, `FN11`, `FN30`, `FN34`.

This is the output of it on Arch Linux, with a working trackpad and trackpoint (also ran `lsdev` and `lsmod`:
```console
$ lsirq | grep rmi
136  104039 dummy 44 rmi4_smbus
139  103934 rmi4 2 rmi4-00.fn11
141     334 rmi4 5 rmi4-00.fn30
137       0 rmi4 0 rmi4-00.fn34
138       0 rmi4 1 rmi4-00.fn01
140       0 rmi4 3 rmi4-00.fn11

$ lsdev | grep rmi
 0  rmi4-00.fn34        137 
 1  rmi4-00.fn01        138 
 2  rmi4-00.fn11        139 
 3  rmi4-00.fn11        140 
 5  rmi4-00.fn30        141 
44  rmi4_smbus          136 

$ lsmod | grep rmi
rmi_smbus              16384  0
rmi_core               86016  1 rmi_smbus
serio                  28672  11 rmi_core,serio_raw,atkbd,hp_accel,psmouse,i8042
```

**Comment by averycblack** [comment]
Hey, I've gotten around to working on adding multiplexing support.
Are you still willing to help out? I realize it's been a few months.
If so, what macOS version are you on? Can you try the below binary please?
Make sure that you have VoodooInput, VoodooPS2Controller, PS2Keyboard, PS2Trackpad, *and* PS2Mouse all enabled.
VoodooRMI should be disabled for now initially. Once I know that the trackpoint and trackpad are detected correctly, with the right behavior, we can enable it then.
Please send a log using `msgbuf=1048576` and `sudo dmesg | grep -i PS2 > ~/Desktop/ps2_log.txt`
An ioreg file from ioregistryexplorer would help too.

Do note that while I don't expect panics, there is still the possibility of panics. Please have some way to recover in case you aren't able to boot with this.
[VoodooPS2Controller-2.2.5-DEBUG.zip](https://github.com/VoodooSMBus/VoodooSMBus/files/6987196/VoodooPS2Controller-2.2.5-DEBUG.zip)

Edit:
If the trackpoint doesn't work, do you mind using the version of VoodooPS2 from acidanthera's repo release page. Check in ioregistryexplorer the value of "RM,TrackpadInfo" for each ApplePS2SynapticTouchpad service. If one of them has type 0x47 and the other type 0x46, try the below binary.
[VoodooPS2Controller-2.2.5-DEBUG.zip](https://github.com/VoodooSMBus/VoodooSMBus/files/6987524/VoodooPS2Controller-2.2.5-DEBUG.zip)

**Comment by XXCoreRangerX** [comment]
Thank you for your work and time with this issue.
I am looking forward to help out, of course.

**Current specs:**
- macOS Big Sur 11.5.2
- OpenCore 0.7.0

---

**Here are the results of my tests:**

Everything went pretty smooth without any panics or crashes.

**1st kext:**
| Device | Working | Buttons | Gestures |
| --- | --- | --- | --- |
| **Trackpoint** | No | No | - |
| **Touchpad** | Yes | No | No |

IOReg and log:
[1st_kext.zip](https://github.com/VoodooSMBus/VoodooSMBus/files/6993868/1st_kext.zip)


**2nd kext:**
| Device | Working | Buttons | Gestures |
| --- | --- | --- | --- |
| **Trackpoint** | Yes | Yes | - |
| **Touchpad** | Yes | No | Partially (scrolling and double tap) |

IOReg and log:
[2nd_kext.zip](https://github.com/VoodooSMBus/VoodooSMBus/files/6993883/2nd_kext.zip)


**Acidanthera’s kext:**
| Device | Working | Buttons | Gestures |
| --- | --- | --- | --- |
| **Trackpoint** | Yes | Yes | - |
| **Touchpad** | Yes | Yes | No |


With the 2nd kext enabled `VoodooPS2Mouse` no longer breaks `VoodooPS2Touchpad`'s support for the touchpad, which is great.

Overall, the 2nd kext works the smoothest out of all but the buttons below the touchpad don't work.

**Comment by averycblack** [comment]
I wonder if the issues with buttons is related to the issues here: https://github.com/acidanthera/bugtracker/issues/1579

The 2nd one indeed does seem to work the best. How well does it work if you add VoodooRMi on top? Keep all the PS2 kexts enabled, and use the VoodooInput from VoodooRMI.

**Comment by XXCoreRangerX** [comment]
This works great! In fact, everything works now!

**2nd kext (VoodooRMI and VoodooSMBus kexts enabled):**
| Device | Working | Buttons | Gestures |
| --- | --- | --- | --- |
| **Trackpoint** | Yes | Yes | - |
| **Touchpad** | Yes | Yes | Yes |

**Edit:**
IOReg and log:
[2nd kext with voodoormi.zip](https://github.com/VoodooSMBus/VoodooSMBus/files/7008846/2nd.kext.with.voodoormi.zip)
